File: javascript/minesweeper/minesweeper.js
function* neighbours(row, col, board) {
    const numRows = board.length;
    const numCols = numRows > 0 ? board[0].length : 0;
    for (let r = row - 1; r <= row + 1; r++) {
        for (let c = col - 1; c <= col + 1; c++) {
            if ((r != row || c != col) && r >= 0 && r < numRows && c >= 0 && c < numCols) {
                yield board[r][c];
            }
        }
    }
}

export function annotate(board) {
    return board.map((row, rowIdx) => {
        return Array.from(row).map((c, colIdx) => {
            if (c == ' ') {
                const mineCount = [...neighbours(rowIdx, colIdx, board)].filter((c) => c == '*')
                                                                        .reduce((prev, _) => prev + 1, 0);
                return mineCount > 0 ? mineCount : ' ';
            } else if (c == '*') {
                return '*';
            } else {
                throw new Error('Bad input');
            }
        }).join('');
    });
}

File: clojure/minesweeper/src/minesweeper.clj
(ns minesweeper
  (:require [clojure.string :as string]))

(def ^:private line-separator (System/getProperty "line.separator"))

(def ^:private neighbour-diffs
  [[-1 -1] [0 -1] [1 -1]
   [-1 0]         [1 0]
   [-1 1]  [0 1]  [1 1]])

(defn draw [board-s]
  (let [board (string/split-lines board-s)
        neighbours-count (fn [pt c]
                           (if (= c \*)
                             "*"
                             (->> neighbour-diffs
                                  (map (partial mapv + pt))
                                  (map (fn [[row col]] (get (get board row) col)))
                                  (filter (partial = \*))
                                  count
                                  (#(if (zero? %) " " (String/valueOf %))))))
        counted-board (map-indexed (fn [row-idx row]
                                     (map-indexed (fn [col-idx c]
                                                    (neighbours-count [row-idx col-idx] c))
                                                  row))
                                   board)]
    (string/join line-separator (map string/join counted-board))))
